src/source/Response.c

```diff
@@ -93,12 +97,25 @@ STATUS freeCurlResponse(PCurlResponse* ppCurlResponse)
         pCallbacksProvider = pCurlResponse->pCurlRequest->pCurlApiCallbacks->pCallbacksProvider;
     }
 
-    if (pCallbacksProvider != NULL && IS_VALID_MUTEX_VALUE(pCurlResponse->lock)) {
-        locked = pCallbacksProvider->clientCallbacks.tryLockMutexFn(pCallbacksProvider->clientCallbacks.customData, pCurlResponse->lock);
-        // if it fail to acquire the lock that means another thread is tearing down the response already.
-        // thus current thread can return early.
-        // TODO: eliminate the possibility of two threads calling freeCurlResponse.
-        CHK(locked, retStatus);
+    if (pCallbacksProvider != NULL) {
+        if (IS_VALID_MUTEX_VALUE(pCurlResponse->lock)) {
+            locked = pCallbacksProvider->clientCallbacks.tryLockMutexFn(pCallbacksProvider->clientCallbacks.customData, pCurlResponse->lock);
+            // if it fail to acquire the lock that means another thread is tearing down the response already.
+            // thus current thread can return early.
+            // TODO: eliminate the possibility of two threads calling freeCurlResponse.
+            CHK(locked, retStatus);
+        }
+        if (IS_VALID_MUTEX_VALUE(pCurlResponse->dataAvailableLock)) {
+            dataAvailableLocked = pCallbacksProvider->clientCallbacks.tryLockMutexFn(pCallbacksProvider->clientCallbacks.customData, pCurlResponse->dataAvailableLock);
+            CHK(dataAvailableLocked, retStatus);

```

Comment:
Need some comments here. So, if we were unable to lock it then does it mean that another teardown is in progress or it means that there is an ongoing operation? In case of the latter, we won't properly teardown the response object (or is it)

Reply:
Perhaps, here, you could trigger an exit with the condition variable and later, when freeing the lock, you could also free the condition variable. NOTE: If you create the condition variable before the data lock, you won't need to check whether its value is valid here.

Think if you need to signal/broadcast termination and await for the condition variable as you hold the lock (try data lock succeeds)